File: contestC#/contest/contest/Repository/AbstractRepository.cs
using contest.Domain;
using contest.Exceptions;
using contest.Utils.Observer;
using contest.Validation;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace contest.Repository
{
    /*
     * Abstract repository class 
     */
    public class AbstractRepository<E, ID> : AbstractObservable, IRepository<E, ID> where E : HasID<ID>
                                                                                       where ID : IComparable<ID>
    {
        /*
         * List of save entities 
         */
        private List<E> entities;

        /*
         * Entity validator
         */ 
        private Validator<E> validator;

        /*
         * Constructor
         */
         public AbstractRepository(Validator<E> validator)
        {
            this.validator = validator;
            entities = new List<E>();
        }

        /*
         * Saves an entity to repository
         * Throws ValidatorException if entity is not valid
         * Throws RepositoryException if it contains a duplicate id 
         */
        public virtual void add(E entity)
        {
            if (getById(entity.Id) != null)
            {
                throw new RepositoryException("This id is already in Repository.");
            }
            validator.validate(entity);
            entities.Add(entity);
            notifyObservers();
        }

        /*
         * Deletes an entity from repository
         * Does not do anything if id is not in repository
         */
        public virtual void delete(ID id)
        {
            E entity = getById(id);
            if (entity != null)
            {
                entities.Remove(entity);
                notifyObservers();
                notifyPushObservers(entity);
            }
        }

        /*
         * Returns all entities
         */ 
        public List<E> getAll()
        {
            return entities;
        }

        /*
         * Returns the entity with the given id 
         * Returns null if the entity is not found
         */
        public E getById(ID id)
        {
            int index = entities.FindIndex(x => x.Id.Equals(id));
            if (index != -1)
            {
                return entities[index];
            } else
            {
                return default(E);
            }
        }

        /*
         * Updates an entity in repository
         * Throws ValidatorException if newEntity is not valid 
         * Throws RepositoryException if original id is not found 
         * Throws RepositoryException if original id is different from new id 
         */
        public virtual void update(ID id, E newEntity)
        {
            validator.validate(newEntity);

            if (!(id.Equals(newEntity.Id)))
            {
                throw new RepositoryException("Id can not be modified.");
            }

            int index = entities.FindIndex(x => x.Id.Equals(id));
            if (index == -1)
            {
                throw new RepositoryException("The original id is not valid.");
            }

            entities[index] = newEntity;
            notifyObservers();
        }

    }
}

File: contestJava/contest/src/Repository/AbstractRepository.java
package Repository;

import Domain.HasId;
import Utils.AbstractObservable;
import Utils.Observer;
import Validator.IValidator;
import Validator.RepositoryException;
import Validator.ValidatorException;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by Sebi on 10/16/2016.
 */
public abstract class AbstractRepository<E extends HasId<ID>, ID extends Comparable<ID>> extends AbstractObservable<E> implements IRepository<E, ID> {
    /*
    Entity validator
     */
    private IValidator<E> validator;

    /*
    Constructor - gets a validator
     */
    public AbstractRepository(IValidator<E> validator){
        this.validator = validator;
    }

    /*
    List of entities from repository
     */
    protected List<E> entities = new ArrayList<E>();

    /*
    Returns the number of entities in repository
     */
    @Override
    public int size(){
        return entities.size();
    }

    /*
    Saves an entity in repository
     */
    @Override
    public void save(E entity) throws RepositoryException, ValidatorException {
        for (E ent : entities){
            if (ent.getId().equals(entity.getId())){
                throw new RepositoryException("This id is already in repository");
            }
        }
        validator.validate(entity);
        entities.add(entity);
        notifyObservers();
    }

    /*
    Updates an entity in repository
     */
    @Override
    public void update(ID id, E newEntity) throws RepositoryException, ValidatorException {
        if (entities.stream().filter(x -> x.getId().equals(id)).count() != 1){
            throw new RepositoryException("The initial id is not valid");
        }
        if (!id.equals(newEntity.getId())){
            throw new RepositoryException("You can't change the id");
        }
        validator.validate(newEntity);
        entities.removeIf(x -> x.getId().equals(id));
        entities.add(newEntity);
        notifyObservers();
    }

    /*
    Deletes an entity from repository
     */
    @Override
    public void delete(ID id) throws RepositoryException {
        long idCount = entities.stream().filter(x -> x.getId().equals(id)).count();
        if (idCount == 0) {
            throw new RepositoryException("This id is not valid");
        }
        entities.removeIf(e -> e.getId().equals(id));
        notifyObservers();
    }

    /*
    Returns the entity with the given id, if found in repository
     */
    @Override
    public E getById(ID id){
        for (E entity : entities){
            if (entity.getId().equals(id)){
                return entity;
            }
        }
        return null;
    }

    /*
    Returns all entities from repository
     */
    @Override
    public List<E> getAll(){
        return entities;
    }

    /*
    Removes all entities from repository
     */
    @Override
    public void clearAll(){
        entities.clear();
        notifyObservers();
    }

}
